### PR TITLE
BUG: linalg.norm(): Don't convert object arrays to float

### LINKED ISSUE #7575: Regression in linalg.norm() using dtype=object
In NumPy 1.10.1, this works:

```python
>>> import numpy as np
>>> np.linalg.norm(np.array([np.array([0, 1]), 0, 0], dtype=object))
array([ 0.,  1.])
```

In NumPy 1.11.0, however, it raises an exception:

```python
>>> import numpy as np
>>> np.linalg.norm(np.array([np.array([0, 1]), 0, 0], dtype=object))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3/dist-packages/numpy/linalg/linalg.py", line 2116, in norm
    x = x.astype(float)
ValueError: setting an array element with a sequence.
```

Probably this

```python
if not issubclass(x.dtype.type, inexact):
    x = x.astype(float)
```

should be changed to

```python
if not issubclass(x.dtype.type, (inexact, object_)):
    x = x.astype(float)
```

?

Does that make sense?

If yes, I can make a PR.

### PR DESCRIPTION
**Pull request by mgeier**
This fixes #7575 (Regression in linalg.norm() using dtype=object).

Please tell me if I shall rebase this onto another branch.

### CODE REVIEW

File: numpy/linalg/tests/test_regression.py [hunk 1]
```diff
@@ -90,6 +90,12 @@ def test_svd_no_uv(self):
                 assert_equal(np.linalg.matrix_rank(a), 1)
                 assert_array_less(1, np.linalg.norm(a, ord=2))
 
+    def test_norm_object_array(self):
```
Comment (charris) [inline numpy/linalg/tests/test_regression.py]: It would be good to test more of the norms, or at least all that work with this fix.

**Comment by charris** [comment]
I've added a comment to the issue #7575.

File: numpy/linalg/tests/test_regression.py (hunk 1 above)
Comment (mgeier) [inline numpy/linalg/tests/test_regression.py]: I've added more tests in bfa3ed9d29314d9d4a06eb5172ae855d19ae7c8d.

Everything that was broken before the regression is still broken.

**Comment by charris** [comment]
I might put this in 1.11.1, but not 1.12,  to give you more time to make your own version of norm and get that out there. Deal?

**Comment by mgeier** [comment]
If it's definitely going to be dropped for 1.12, I don't care about what's happening in 1.11.1.
I could very quickly switch to the work-around given by @pv in #7575.

But I don't think this would be a good decision for the `linalg` submodule to demote object arrays from being first-class citizens.

Let me offer a counter-deal:
You add my fix to 1.11.1, in the meantime I'll try to come up with a PR that adds support for object arrays in `numpy.linalg.norm()` for all possible values of the `ord` keyword.
If you are satisfied with my PR, it goes into 1.12. If not (or if I can't come up with one until the deadline for 1.12), you can drop the whole thing (and we are back to your deal from above).

What about that?

**Comment by charris** [comment]
Let's put this in with the understanding that it is something of an implementation detail. If you want a guarantee that your work will not be affected by future changes you should pursue your own implementation of a norm function that works as you need it to for your application.

**Comment by charris** [comment]
Thanks @mgeier .

**Pull request merged**